pinot-core/src/main/java/org/apache/pinot/core/transport/server/routing/stats/ServerRoutingStatsManager.java

```diff
@@ -0,0 +1,373 @@
+    if (!_isEnabled) {
+      return null;
+    }
+
+    ServerRoutingStatsEntry stats = _serverQueryStatsMap.get(server);
+    if (stats == null) {
+      return null;
+    }
+
+    try {
+      stats.getServerReadLock().lock();
+      return stats.getNumInFlightRequests();
+    } finally {
+      stats.getServerReadLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a list containing each server and the corresponding EMA latency seen for queries on the server.
+   */
+  public List<Pair<String, Double>> fetchEMALatencyForAllServers() {
+    List<Pair<String, Double>> response = new ArrayList<>();
+    if (!_isEnabled) {
+      return response;
+    }
+
+    for (Map.Entry<String, ServerRoutingStatsEntry> entry : _serverQueryStatsMap.entrySet()) {
+      String server = entry.getKey();
+      Preconditions.checkState(entry.getValue() != null, "Server stats is null");
+      ServerRoutingStatsEntry stats = entry.getValue();
+
+      stats.getServerReadLock().lock();
+      double latency = stats.getLatencyEMA();
+      stats.getServerReadLock().unlock();
+
+      response.add(new ImmutablePair<>(server, latency));
+    }
+
+    return response;
+  }
+
+  /**
+   * Same as above but returns the EMA latency for the input server.
+   */
+  public Double fetchEMALatencyForServer(String server) {
+    if (!_isEnabled) {
+      return null;
+    }
+
+    ServerRoutingStatsEntry stats = _serverQueryStatsMap.get(server);
+    if (stats == null) {
+      return null;
+    }
+
+    try {
+      stats.getServerReadLock().lock();
+      return stats.getLatencyEMA();
+    } finally {
+      stats.getServerReadLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a list containing each server and the corresponding Hybrid score for each server. The Hybrid score is
+   * calculated based on https://www.usenix.org/system/files/conference/nsdi15/nsdi15-paper-suresh.pdf.
+   */
+  public List<Pair<String, Double>> fetchHybridScoreForAllServers() {
+    List<Pair<String, Double>> response = new ArrayList<>();
+    if (!_isEnabled) {
+      return response;
+    }
+
+    for (Map.Entry<String, ServerRoutingStatsEntry> entry : _serverQueryStatsMap.entrySet()) {
+      String server = entry.getKey();
+      Preconditions.checkState(entry.getValue() != null, "Server stats is null");
+      ServerRoutingStatsEntry stats = entry.getValue();
+
+      stats.getServerReadLock().lock();
+      double score = stats.computeHybridScore();
+      stats.getServerReadLock().unlock();
+
+      response.add(new ImmutablePair<>(server, score));
+    }
+
+    return response;
+  }
+
+  /**
+   * Same as above but returns the score for a single server.
+   */
+  public Double fetchHybridScoreForServer(String server) {
+    if (!_isEnabled) {
+      return null;
+    }
+
+    ServerRoutingStatsEntry stats = _serverQueryStatsMap.get(server);
+    if (stats == null) {
+      return null;
+    }
+
+    try {
+      stats.getServerReadLock().lock();
+      return stats.computeHybridScore();
```

Comment:
This is fine. `readLock` is needed here

Reply:
Ack. Will be addressed by TODO in subsequent PR.